### PR TITLE
[WIP] Vuex Store Blockchain aepp-sdk Implementation

### PR DESCRIPTION
**Pull request by sqevani**
### Task
_Vuex Store Blockchain aepp-sdk Implementation_

### Description
_Due to the constant change of the REST API of the aeternity blockchain, we need to implement the SDK internally for the blockchain explorer, to minimize the need to continuously update the app and waste time_

### Flow
_Currently I'll try to refactor the Vuex-Store and implement it so its modular, and context specific, the idea is that the Vuex-Store will act as a middleware for the SDK and the Vue App (Blockchain explorer)_

* Refactor Vuex-Store into modular parts
* Reading on the AEPP-SDK API
* Implement AEPP-SDK API into Vuex Actions
* Bind data to the views

__PS: WIP stands for Work in Progress__

### CODE REVIEW

**Comment by sadiqevani** [comment]
> I don't understand how this store refactoring can minimize the need to update the app in case of Epoch breaking changes. Basically, I don't agree with changes proposed in #26. I think we just need to replace fetchJson with calls to aepp-sdk in this version of blockchain explorer store. And update these calls in future if aepp-sdk will have breaking changes. So, "internal aepp-sdk" was already implemented there because components don't directly depend on Epoch api and all breaking changes on aepp-sdk can be fixed in that single file.

While I was in Berlin, @dadaphl told me that long-term we'll be using the AEPP-SDK in all the applications, so, integrating the AEPP-SDK into the VueAPP is a long-term decision, so we can simply update the code in the store-actions in case things change.

> Why are you adding so lot of obvious comments in the code? Who except you will maintain them?

Code is cleaner, and I'm not going to be the only developer looking at it and knowing what it does. For me and you might be obvious, but sometimes not so obvious for other developers.

This is an open source project, and the cleaner the things look, the easier is for external developers to understand what's going on.

Also, I usually try to use a small set of features of a library/framework/language, since you need 20% to complete 80%, and reusing the same structure, makes it easier to come back and look at the code and remember it easier.

So, instead of using the "cleverest tricks" I would like to use the most boring of options, so things are tight looking, clean, and easy to understand.

**Comment by sadiqevani** [comment]
> Who except you will maintain them?

Generally the way the comments are laid down are simple one liners that don't need much updates, in case something changes, there's a nigh chance, that the whole code with the comment will go away, and the comment need to be rewritten.

So I don't think there's going to be a problem updating that one-line comment.

**Comment by dadaphl** [comment]
> Basically, I don't agree with changes proposed in #26.

I do. It was a task I specifically asked for to be completed. This application is growing and it makes sense to split parts.

> So, "internal aepp-sdk" was already implemented ...

Sorry, I don't understand

> @dadaphl told me that long-term we'll be using the AEPP-SDK in all the applications

correct

**Comment by davidyuk** [comment]
> I do. It was a task I specifically asked for to be completed. This application is growing and it makes sense to split parts.

Of course, the app will grow if we add 4 lines of comments (empty line, two lines with brackets, and actual comment) for every line of code.

> I do. It was a task I specifically asked for to be completed. This application is growing and it makes sense to split parts.

> @dadaphl told me that long-term we'll be using the AEPP-SDK in all the applications

so, we are going to reuse store or it's modules of blockchain explorer in other applications?

>> So, "internal aepp-sdk" was already implemented ...

> Sorry, I don't understand

I mean "implement the SDK internally for the blockchain explorer" from the discription of this PR, it was already implemented.

**Comment by sadiqevani** [comment]
> Of course, the app will grow if we add 4 lines of comments (empty line, two lines with brackets, and actual comment) for every line of code.

Well the way I see it, is that for me clarity comes before anything else. Given clarity, any problem can be solved, if a problem is unclear, then solutions will be harder to come by.

Now, also for me there is a balance between clean code and also clear code, by clear code i mean visually pleasing, so you can go from top/bottom without getting overwhelmed by all functions/variables crammed into a single line of code, where to come to a conclusion does 1001 steps.

Our brain generally can keep up to 4 concepts in the focus mode, thats a hard limit, so I usually like to split actions with comments, and go from top->bottom, so the brain does not have to keep multiple variables/functions/results in mind when going through the code.

> so, we are going to reuse store or it's modules of blockchain explorer in other applications?

It can be reused, but for now I don't think that will be the case.

> I mean "implement the SDK internally for the blockchain explorer" from the discription of this PR, it was already implemented.

I'm not reimplementing the SDK, i'm integrating the SDK with Vuex-store, its an interface for Vue to access. (poor choice of words on my side)

**Comment by dadaphl** [comment]
> > > > So, "internal aepp-sdk" was already implemented ...

> > > Sorry, I don't understand

> > I mean "implement the SDK internally for the blockchain explorer" from the discription of this PR, it was already implemented.

> I'm not reimplementing the SDK, i'm integrating the SDK with Vuex-store, its an interface for Vue to access. (poor choice of words on my side)

Yes, there seems to be a misunderstanding. The task was to integrate the aepp-sdk. Recently api endpoints keep changing over and over again. It is expected that due to adoption of bitcoin-ng, that the api will change once again. Therefore the aepp-sdk parses swagger definition provided by the api and determines the current endpoint. As long as the features provided by the api and the return data don't change radically, the blockchain explorer should work with multiple versions of the epoch api.

**AGAIN: "integration of the aepp-sdk" is a task I SPECIFICALLY ASSIGNED to @sqevani** 

@davidyuk If you disagree with this Idea in general, you should direct your concerns to @emil-apeunit or me. 

I appreciate a lot that you contribute and engage in the discussion. **But:**

> Of course, the app will grow if we add 4 lines of comments (empty line, two lines with brackets, and actual comment) for every line of code.

I don't think that sarcasm will help at all in a healthy discussion. This is NOT constructive criticism. You might be right, or not. But you have to find a different tune to address your concerns and help guide development in the right direction.

File: src/components/blank/style.scss [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+@mixin prefix($property, $value, $prefixes: ()) {
+  @each $prefix in $prefixes {
+    -#{$prefix}-#{$property}: $value;
+  }
+
+  #{$property}: $value;
+}
+
+.tag-blank {
+  @include prefix(transform, translate(-50%, -50%), 'webkit' 'ms');
```
Comment (davidyuk) [inline src/components/blank/style.scss]: Can't this be handled by `autoprefixer`?

Comment (sadiqevani) [reply src/components/blank/style.scss]: its temporary, will remove it (just trying some stuff out)

File: src/store/modules/accounts/actions.js [hunk 2]
```diff
@@ -15,31 +18,46 @@ const { startLoading, endLoading } = createActionHelpers({
...
-       */
-      return resolve('template')
-    })
+  async get ({ state, commit, dispatch }, address) {
+    startLoading(dispatch, 'accounts/get')
+
+    const client = await ae
+
+    const [{ balance }, { transactions }] = await Promise.all([
+      await client.api.getAccountBalance(address),
```
Comment (davidyuk) [inline src/store/modules/accounts/actions.js]: If you are using `await` here, it means that `Promise.all` will receive an array of actual values instead of promises, so `Promise.all` has no sense in this case.

File: src/store/modules/accounts/actions.js [hunk 3]
```diff
@@ -15,31 +18,46 @@ const { startLoading, endLoading } = createActionHelpers({
...
+    commit('setAccount', account)
+
+    return endLoading(dispatch, 'accounts/get')
+  },
+
+  async name ({ state, commit, dispatch }, address) {
+    startLoading(dispatch, 'accounts/name')
+
+    if (!process.env.NAME_LOOKUP_MIDDLEWARE_URL) return
+    if (!isEmpty(state.name)) if (new Date().getTime() - state.name.ts < 10000) return
```
Comment (davidyuk) [inline src/store/modules/accounts/actions.js]: You can use `Date.now()` instead of `new Date().getTime()`

File: src/store/modules/blocks/actions.js [hunk 4]
```diff
@@ -15,36 +18,105 @@ const { startLoading, endLoading } = createActionHelpers({
...
+    const client = await ae
+    const block = await client.api.getBlockByHeight(height, { txEncoding: 'json' })
+
+    if (isEqual(state.block, block)) {
+      endLoading(dispatch, 'blocks/getBlockFromHeight')
+      return state.block
+    }
+
+    commit('setBlock', block)
+    endLoading(dispatch, 'blocks/getBlockFromHeight')
```
Comment (davidyuk) [inline src/store/modules/blocks/actions.js]: I think `startLoading(dispatch, <method name>)` and `endLoading(dispatch, <method name>)` should be extracted to decorator to avoid duplication of this pattern.

**Comment by davidyuk** [comment]
> Therefore the aepp-sdk parses swagger definition provided by the api and determines the current endpoint. As long as the features provided by the api and the return data don't change radically, the blockchain explorer should work with multiple versions of the epoch api.

`client.api.*` methods of `aepp-sdk` are generated by swagger definition, so changes in swagger (Epoch api) will break Blockchain explorer with the same probability as without `aepp-sdk`. But, in general, to use `aepp-sdk` is better.

File: src/store/modules/accounts/actions.js (hunk 2 above)
Comment (sadiqevani) [reply src/store/modules/accounts/actions.js]: Will update this. Though this was legacy code, that i copy/pasted to have an idea of what to get from it.

File: src/store/modules/accounts/actions.js (hunk 3 above)
Comment (sadiqevani) [reply src/store/modules/accounts/actions.js]: Will update.

File: src/store/modules/accounts/mutations.js [hunk 5]
```diff
@@ -3,12 +3,21 @@
...
+    return Object.assign(state, { account })
+  },
+
+  /**
+   * setName
+   * @param {Object} state
+   * @param {String} name
+   */
+  setName: function (state, name) {
+    return Object.assign(state, { name })
```
[the remark this answers is not in the document]
Comment (sadiqevani) [reply src/store/modules/accounts/mutations.js]: Just a personal preference, some edge case might need it, and I don't see any issue in returning. in FP thats the core functionality of a function that has no side effects. (though this case its not FP, but either way still stands)

(personal dogma, always return something from a function)

File: src/store/modules/blocks/actions.js [hunk 6]
```diff
@@ -15,36 +18,105 @@ const { startLoading, endLoading } = createActionHelpers({
...
+   * @param {String} hash
+   * @return {*}
+   */
+  async getBlockFromHash ({ state, commit, dispatch }, hash) {
+    startLoading(dispatch, 'blocks/getBlockFromHash')
+
+    const client = await ae
+    const block = await client.api.getBlockByHash(hash, { txEncoding: 'json' })
+
+    if (isEqual(state.block, block)) {
```
[the remark this answers is not in the document]
Comment (sadiqevani) [reply src/store/modules/blocks/actions.js]: I see your point, but this function returns two different states, which then can be used to continue some kinda of flow in the vue components.

Let say if you want to chain multiple actions based on the previous value of the state, without recurring to directly access the store through `this.$store`

[the remark this answers is not in the document]
Comment (davidyuk) [reply src/store/modules/blocks/actions.js]: two different states? As I see, the first one is `block` from state and the second one is `block` received from `aepp-sdk`. The difference is that `block` from the state is reactive variable and the second are not. For consistency will be better to return always a reactive version.
As I know, `store.commit` runs synchronously, so it is possible to use just:
```js
if (!isEqual(state.block, block)) {
  commit('setBlock', block)
}
endLoading(dispatch, 'blocks/getBlockFromHash')
return state.block
```

**Comment by dadaphl** [comment]
> changes in swagger (Epoch api) will break Blockchain explorer with the same probability as without aepp-sdk

not true

The core team and aepp-dev-tools team agreed that `operationId` in swagger will not change. The URLs will change. Aepp-sdk maps oprationId to URL.

Of course that doesn't help if the underlying API changes radically.

**Comment by dadaphl** [comment]
@sadiqevani there is a spinner calld ae-loader already in the components
https://components.aepps.com/#aeloader

Please create PRs that do one thing at a time. The title of the PR is 'Vuex Store Blockchain aepp-sdk Implementation'. I'd like to see the PR only focus on that. Adding a spinner and loading animation should be a different PR altogether. For now keep it that way because I don't want to waste time on refactoring PRs, but please keep that in mind for the next time.

File: src/screens/status/status.vue [hunk 7]
```diff
@@ -48,12 +48,12 @@
...
 import pollAction from '@/mixins/pollAction'
 export default {
-  mixins: [pollAction('fetchNodeStatus')],
+  mixins: [pollAction('getNodeStatus')],
   data: () => ({
     node: process.env.AETERNITY_EPOCH_API_URL
   }),
   computed: mapState({
-    s: 'nodeStatus'
+    s: '$nodeStatus'
```
Comment (davidyuk) [inline src/screens/status/status.vue]: Have you explain what `$` means somewhere?

Comment (sadiqevani) [reply src/screens/status/status.vue]: This is the syntax for the store:

- `${name of the variable}` -> global property unto the state
- `_{name of the module}` -> partials/component state, ex: modal, tooltip, etc..
- `{everything else}` -> modules

Reduces name collisions, makes it more clear where it comes from, though need to write this somewhere/readme.md

**Comment by sadiqevani** [comment]
@dadaphl 

I managed to make the homepage + blocks work, kinda slight refactoring, though what's left are some small parts with the `transactions`, from what I saw from the API there's no specific api endpoint to get transactions?

We have to pull that from the `block.transations` array, right?

Also for the `$nodeStatus` the call here:

```
 /**
     * Pull node state, version and peers
     */
    const [nodeTop, nodeVersion, peers] = await Promise.all([
      fetchJson(`${process.env.AETERNITY_EPOCH_API_URL}v2/top`),
      fetchJson(`${process.env.AETERNITY_EPOCH_API_URL}v2/version`),
      Promise.all(times(3, async (i) => {
        const [[top, address], version] = await Promise.all([
          fetch(`${process.env.AETERNITY_EPOCH_API_URL}peer/${i + 1}/v2/top`).then(async response => {
            return [await response.json(), response.headers.get('X-Upstream')]
          }),
          fetchJson(`${process.env.AETERNITY_EPOCH_API_URL}peer/${i + 1}/v2/version`)
        ])
        return { top, version, address }
      }))
    ])
```

When its connecting to the peers, the endpoint does not work, it changed? replaced? removed?

File: src/screens/status/status.vue (hunk 7 above)
Comment (dadaphl) [reply src/screens/status/status.vue]: Yes, if you introduce conventions, please document them.

File: src/partials/marketStats/marketStats.vue [hunk 8]
```diff
@@ -34,31 +34,35 @@
...
+
+  /*
+   * Create poll function for market-stats
+   */
+  mounted: function () {
+    return poll.fetch.call(this, '_marketStats/get')
+  },
+  destroyed: function () {
+    return poll.close('blocks/getLatestBlocks')
+  }
```
Comment (davidyuk) [inline src/partials/marketStats/marketStats.vue]: `pollAction` mixin was implemented actually to avoid this pattern
https://github.com/aeternity/aepp-blockchain-explorer/pull/10#issuecomment-379461890

File: src/store/modules/blocks/getters.js [hunk 9]
```diff
@@ -1,4 +1,31 @@
...
+   * getAverageBlockTime by calculating
+   * the list of the blocks on the state
+   * @param state
+   * @return {number}
+   */
+  getAverageBlockTime: function (state) {
+    const blockTimes = state.blocks.map(
+      (block, idx, blocks) => idx + 1 < blocks.length && blocks[idx].time - blocks[idx + 1].time
+    )
+    // no idea what this does
```
Comment (davidyuk) [inline src/store/modules/blocks/getters.js]: it creates a new array every element of it is a difference between creation times of two corresponding blocks

File: src/router/index.js [hunk 10]
```diff
@@ -12,44 +12,36 @@ import Status from '@/screens/status/status.vue'
...
+  }, {
+    name: 'TransactionDetail',
+    path: '/tx/:txId',
+    component: TransactionDetail,
+    props: true
+  }, {
+    name: 'Status',
+    path: '/status',
+    component: Status
+  }]
```
Comment (davidyuk) [inline src/router/index.js]: Is this change related to store refactoring?

File: src/store/modules/transactions/actions.js [hunk 11]
```diff
@@ -25,8 +26,81 @@ export default {
...
+   * @param {Object} state
+   * @param {Function} commit
+   * @param {Function} dispatch
+   * @param {String} hash
+   * @return {*}
+   */
+  async mempool ({ state, commit, dispatch }, hash) {
+    startLoading(dispatch, 'transactions/mempool')
+
+    const client = await ae
```
Comment (davidyuk) [inline src/store/modules/transactions/actions.js]: Is it possible to put an initialized client into the state, to avoid `const client = await ae` duplication?

File: src/functions/polling.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    },
+
+    /**
+     * close function clears the interval
+     * for the specific poll action
+     * @param action
+     * @param callback
+     * @return {*}
+     */
+    close (action, callback) {
```
Comment (davidyuk) [inline src/functions/polling.js]: This function shouldn't receive a callback because it runs synchronously.

File: src/screens/blockList/blockList.vue [hunk 13]
```diff
@@ -115,6 +114,22 @@ export default {
     loadMore: function () {
       return this.$store.dispatch('blocks/getLatestBlocks', this.blocks.length + 10)
     }
+  },
+
+  /*
+  * Before and After route events
+  */
+  beforeRouteEnter (to, from, next) {
```
Comment (davidyuk) [inline src/screens/blockList/blockList.vue]: Is not easier to use `mounted`/`beforeDestroy`?

File: yarn.lock [hunk 14]
```diff
@@ -13,9 +13,9 @@
     string-loader "0.0.1"
     vue-clickaway "^2.1.0"
 
-"@aeternity/aepp-sdk@github:aeternity/aepp-sdk-js#feature/sweet-api":
-  version "0.14.0-0.1.0-alpha.1"
-  resolved "https://codeload.github.com/aeternity/aepp-sdk-js/tar.gz/8dbad4eb72004115f9bbf1d822b8460a36ad8df1"
+"@aeternity/aepp-sdk@github:aeternity/aepp-sdk-js#develop":
+  version "0.15.0-0.1.0"
+  resolved "https://codeload.github.com/aeternity/aepp-sdk-js/tar.gz/7cda8877e9f69a6584105749febc10ccb7dfbb89"
```
Comment (davidyuk) [inline yarn.lock]: New versions of this packages are related to changes in this commit?

Comment (sadiqevani) [reply yarn.lock]: @davidyuk no. I'm using yarn and npm at the same time due to some issues, with yarn not building dependencies, for the `ae-components`

File: src/store/modules/accounts/mutations.js [hunk 15]
```diff
@@ -9,7 +19,7 @@ export default {
    * @return {Object}
    */
   setAccount: function (state, payload) {
-    return Object.assign(state.accounts, { [payload.address]: payload })
+    return Vue.set(state, 'accounts', { [payload.address]: payload })
```
[the remark this answers is not in the document]
Comment (sadiqevani) [reply src/store/modules/accounts/mutations.js]: @davidyuk will change it.

File: src/store/modules/transactions/actions.js (hunk 11 above)
Comment (sadiqevani) [reply src/store/modules/transactions/actions.js]: @davidyuk okay

File: src/router/index.js (hunk 10 above)
Comment (sadiqevani) [reply src/router/index.js]: @davidyuk no, got committed by accident

**Comment by sqevani** [comment]
@dadaphl so, how do we merge this baby? 👶💃

File: src/functions/polling.js [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+  /*
+   * Setting the interval and returning the id
+   */
+  return setInterval(() => this
+  .$store
+  .dispatch(action, args), ms)
+}
+
+export default function () {
```
Comment (davidyuk) [inline src/functions/polling.js]: This function is actually constructor, so it should be ES6 Class instead.

File: src/store/modules/transactions/actions.js (hunk 11 above)
Comment (davidyuk) [reply src/store/modules/transactions/actions.js]: In the Base app, I'm using [this plugin](https://github.com/aeternity/aepp-identity/blob/2028c64/src/store/plugins/initEpoch.js). In proposed implementation exception will happen if this action will be called before client initialization, but it won't if we will handle polling on the side of Vue store instead of polling function/mixin on the side of Vue components. As I see it, components should maintain a list of subscriptions in a global state (by "subscribe"/"unsubscribe" mutations) and display value of reactive variables from the store. Accordingly, the store should make polling according to a list of subscriptions, and update state displayed by components. Such architecture will make components simpler.

File: src/router/index.js (hunk 10 above)
Comment (davidyuk) [reply src/router/index.js]: You can fix it by rebasing.

**Comment by sadiqevani** [comment]
@dadaphl @davidyuk good to merge?

@davidyuk the improvements you've suggested on the comments, will be handled in another pull-request.

**Comment by dadaphl** [comment]
I completely agree with @davidyuk. I will merge this PR still because we need to make progress. Next time I will not do it.
